### PR TITLE
Implement Set inside Variant

### LINKED ISSUE #867: Implement Hashset primitive to compliment Dictionary & Array
**Describe the project you are working on:**
An old-school match-three game, although I've encountered the need for a Hashset on other projects, and I can foresee its need on future projects.

**Describe the problem or limitation you are having in your project:**
Sometimes what you care about most is the relationship between data, even more so than the data itself. In my case, I care about the _set_ of data I have, and I want to easily perform traditional set operations on it, such as checking if a value exists in my collection of data, or if one set overlaps, is a subset/superset of another set, etc.

My explicit example and what prompted me to open an issue, is that in finding matches for items on my game board, I need to quickly check if an item already belongs to another match that was found, determine if that type of item has already been handled by any matches, and determine if different matches overlap so that I can properly merge them. As well as numerous other little things that require the same functionality.

I'm working around this by using Dictionary and only making use of the keys (just set the values to 0) when doing my checks, but what I'd ideally have access to is a Hashset. A data structure that doesn't store values, only keys, explicitly for performing fast set operations upon.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
This is related to proposal #821, but where it's concerned with adding set operations to existing primitives, I'd like to see an actual Hashset implemented, as it's best suited to situations where you explicitly need the lookup characteristics of a Dictionary, but do not have key:value pairs to store.

As you can see from the conversation in proposal #821, there is some ambiguity in set operations on Dictionaries, whereas there is none on a Hashset, since it stores only keys.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I haven't dug enough into the Godot source to say how Dictionary and other primitives are implemented, but I would imagine it would be much like Dictionary in execution, but with the ability to store values removed and set operation functions as discussed in #821 added.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
A Hashset could be developed in GDScript, but it would be slow and not a primitive like Array or Dictionary. There is no way to work around that in GDScript. For how often it would be used, I would imagine semi-regularly. Many people (including myself) tend to design their code around the [available] data-structure that best fits what you're trying to do.

**Is there a reason why this should be core and not an add-on in the asset library?:**
I don't believe it's possible to add primitive types through GDScript or GDNative, so this can only be added into the core engine itself.

### PR DESCRIPTION
**Pull request by radiantgurl**
closes https://github.com/godotengine/godot-proposals/issues/867

- [x] Core
- [x] Documentation
- [x] Editor
  - [ ] Icon
- [x] GDScript
  - [x] Implementation
  - [ ] Tests
     - [x] Non-typed test
     - [ ] Typed test

TODO:
 - [x] Modify the VariantParser to no longer use ERR_PRINTER_ON_FIRE to signal from the dictionary parser that it's actually a set.
 - [x] Modify for it to support sets with one element.
 - [x] Make the unit tests stop failing even if the output is correct.
 - [x] Implement typed sets
 - [ ] Fix issues that make the typed sets unit test fail.
 
 TODO AFTER MERGE:
 - Mono
 - Make sets exportable.
 - Add set to godot-cpp

### CODE REVIEW

**Comment by AThousandShips** [comment]
Note that the opposite of "disjoint" is "non-disjoint" or better "overlapping", if we're going to use set theory terms I think we should use established ones ("joint" is not one), but I'd say the more accessible and generally useful phrasing would be "non-overlapping" and "overlapping" for what are now called "disjoint" and "joint"

**Comment by RadiantUwU** [comment]
Using placeholder icon for now.

**Comment by fire** [comment]
`ERROR: @GlobalScope.xml: Unrecognized opening tag "[Set]" in method "is_same" description.` Strange error in the cicd.

**Comment by RadiantUwU** [comment]
Using placeholder icon for now.

**Comment by RadiantUwU** [comment]
Woops

**Comment by RadiantUwU** [comment]
image
Doesnt pass tests, i have no idea why. The file output matches with the test.

**Comment by RadiantUwU** [comment]
Since the file output matches the unit tests expected, i won't let it stop me.

**Comment by fire** [comment]
I personally would like set even though it bloats variant by another type. Not sure how much of a review I can do.

File: core/variant/set.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "set.h"
+
+#include "core/templates/hash_set.h"
+#include "core/templates/safe_refcount.h"
+#include "core/variant/variant.h"
+// required in this order by VariantInternal, do not remove this comment.
```
Comment (AThousandShips) [inline core/variant/set.cpp:36]:
```suggestion
```
Unnecessary

File: modules/gdscript/gdscript_parser.h [hunk 2]
```diff
@@ -33,6 +33,7 @@
 
 #include "gdscript_cache.h"
 #include "gdscript_tokenizer.h"
+#include "scene/main/node.h"
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_parser.h:36]:
```suggestion
```
Irrelevant include for the header, probably false positive for the internal `Node`, I'd suggest turning off any IDE feature adding includes as they're usually wrong

**Review by AThousandShips — COMMENTED** [review]
Some initial feedback

File: modules/gdscript/gdscript_parser.cpp [hunk 3]
```diff
@@ -3041,6 +3048,41 @@ GDScriptParser::ExpressionNode *GDScriptParser::parse_dictionary(ExpressionNode
 	return dictionary;
 }
 
+GDScriptParser::ExpressionNode *GDScriptParser::parse_set(ExpressionNode *p_previous_operand, bool p_can_assign) {
+	DictionaryNode *dictionary = (DictionaryNode *)p_previous_operand;
```
Comment (HolonProduction) [inline modules/gdscript/gdscript_parser.cpp:3052]: This makes the unchecked assumption that the passed operand is always a dictionary node. If this is intended make the type of the parameter `DictionaryNode *` instead to get compiler errors if used wrong. If not you will need to check whether the operand is a dictionary node by using `Node.type`

File: modules/gdscript/gdscript_parser.cpp [hunk 4]
```diff
@@ -2975,6 +2975,13 @@ GDScriptParser::ExpressionNode *GDScriptParser::parse_dictionary(ExpressionNode
 					case GDScriptTokenizer::Token::EQUAL:
 						dictionary->style = DictionaryNode::LUA_TABLE;
 						break;
+					case GDScriptTokenizer::Token::COMMA:
```
Comment (HolonProduction) [inline modules/gdscript/gdscript_parser.cpp:2978]: I don't think that `parse_dictionary` should call to `parse_set`. Better use another method `parse_dictioanry_or_set` that delegates to both of the methods.

File: modules/gdscript/gdscript_parser.cpp [hunk 5]
```diff
@@ -3041,6 +3048,41 @@ GDScriptParser::ExpressionNode *GDScriptParser::parse_dictionary(ExpressionNode
...
+
+	if (!check(GDScriptTokenizer::Token::BRACE_CLOSE)) {
+		do {
+			if (check(GDScriptTokenizer::Token::BRACE_CLOSE)) {
+				// Allow for trailing comma.
+				break;
+			}
+
+			// Element.
+			ExpressionNode *element = parse_expression(false, true); // Stop on "=" so we can check for Lua table style.
```
Comment (HolonProduction) [inline modules/gdscript/gdscript_parser.cpp:3070]: We don't want lua table style in sets, so we don't need to stop on "=" I'd say.

File: modules/gdscript/gdscript_analyzer.cpp [hunk 6]
```diff
@@ -3592,6 +3597,23 @@ void GDScriptAnalyzer::reduce_dictionary(GDScriptParser::DictionaryNode *p_dicti
...
 }
 
+void GDScriptAnalyzer::reduce_set(GDScriptParser::SetNode *p_set) {
+	HashSet<GDScriptParser::ExpressionNode *, VariantHasher, StringLikeVariantComparator> elements;
+
+	for (GDScriptParser::ExpressionNode *expr : elements) {
+		reduce_expression(expr);
+	}
+
+	// It's dictionary in any case.
```
Comment (HolonProduction) [inline modules/gdscript/gdscript_analyzer.cpp:3607]: It's a set

**Review by HolonProduction — COMMENTED** [review]
Overall I'm not convinced of the GDScript syntax `{A, B, C}`. It adds unnecessary complexity when parsing. Also I don't think it is very readable. I would prefer to constructs sets like packed arrays:
`Set()`.

**Comment by Calinou** [comment]
I'd say `{A, B, C}` syntax can be confusing, since it can be mistaken for a dictionary at a glance. I would also prefer an explicit `Set()` syntax given this type won't be used very often.

**Comment by HolonProduction** [comment]
Could you elaborate why you think it would be better to understand/comprehend? Not everyone might be familiar with mathematical notations. And while not everyone might be familiar with the term "Set" as well, it at least gives a good search query to start (and a target for symbol lookup to get to the documentation).

Also while the notation is well established in math, I don't think it is in programming, or is it? I personally wouldn't feel certain that I'm looking at a set if I see it in a language that I didn't know.

It might be a tradeoff for the usage if you have to write it out all the time. But given it's only three letters I don't think it's that important.

If it turns out to be a real problem we could implement vararg methods for common set operations, but given that there is no overloading this might get confusing.

**Comment by Mickeon** [comment]
I myself brought up concerns about the name "Set" itself. I'm fully aware this is how it is most commonly known as, but I wonder if it's worth evaluating another name for this type.

Main reason is because "set" is used in Godot exclusively to refer to... a setter, setting a property.
Second reason, and I know the following will sound extremely nitpicky, but a lot of class reference will have to be modified to avoid any possible ambiguity, as using "set" to refer to a group of elements in an array is reasonably common.

**Comment by HolonProduction** [comment]
I mean I can kind of understand that sentiment, but I can't really think of any another name that wouldn't lead to the same effect in the other direction (avoiding calling our `NotSet`s set). Also I don't really think that has been a point of confusion in other languages. E.g. Kotlin has both `Set` type and a `set` keyword.

I mean if there is a concrete idea for an alternative name we can evaluate it, but I can't think of one.

**Comment by Bromeon** [comment]
Agree with Mickeon, it's particularly bad because Godot uses a ton of "set" identifiers all over the place, including `set()` as a method on _every single_ class. Simply calling it `HashSet` would already solve most issues, and it makes also clear that it's O(1) and not a tree-based implementation. This is further clarified for people coming from C++ and being used to `std::set`.


> E.g. Kotlin has both `Set` type and a `set` keyword.

But `set` doesn't usually appear as part of method names, when it refers to "setting a value". In fact, Kotlin is a good example because it entirely replaced setter methods with properties. So the confusion is naturally lower.

**Comment by Bromeon** [comment]
Some thoughts regarding this feature in general:

1. Is `HashSet` conceptually identical to a `Dictionary` with `null` (unit) values, except maybe faster? Or are there notable differences?

1. How would `HashSet` interoperate with other data structures like `Array` or `Dictionary`?
    - How would conversions look?

1. Is there a strategy for dealing with mutations of the key type? 
    - Mostly a problem for objects with custom hashes. 
    - Mutation should likely be disallowed, but maybe there could be Debug checks or so, also useful in combination with iteration (but not sure how hard that would be...)

1. Are there cases where the Godot API itself benefits from this type? Or is it mostly a user convenience?

1. How do users and Godot API developers choose between `Array` and `HashSet`?
    - `Array` is often the faster choice _even_ if you need a set, unless you have a huge number of elements. Linear search on small numbers is very efficient, and you benefit from contiguous storage, especially for iteration.
    - Until now, existing APIs use `Array` even when elements are unique and order is not essential. These can't change for compatibility reasons, even after `HashSet` is added, so we have potential inconsistencies.
    - We need to keep in mind that we had this exact situation with `String` and `StringName` -- adding `StringName` led to many inconsistent APIs, and the implicit conversions caused performance problems in some cases (simply passing strings could lead to expensive conversions in every call). There needs to be a strategy for this.

1. Are there plans for making this typed in the future, i.e. `HashSet[T]`?

**Comment by AThousandShips** [comment]
I think this should be a `RefCounted` type like `PackedDataContainer` instead of a `Variant` type, I don't think the general demand or need for this type is great enough to justify the major overhead and changes to core code. Especially since this does *not* use typing, if `Set[T]` was a thing then this would be far more reasonable as a `Variant` type, but then that would involve even more code and complexity which I don't think is necessary

The main reasons to consider this as a `Variant` type would be, IMO:
1) Support for initializers
2) Typing support
3) Cleaner serializing support
4) Tighter integration with extensions and C#

But 1 has been questioned, and I agree that it both isn't very necessary to have it, and that it would be hard to do so in a way that doesn't both create confusion and make parsing more complicated

For 2 there's no implementation for it, and doesn't seem to be any intention to add it

I also don't necessarily see the need for 3, I'd consider sets here a matter of processing and not data storage, and with a good set of methods you can easily just input an `Array` of data into the set as part of some de-serialization method if you desire

So for me only 4 stands out as a real benefit here, but I don't think a high performance and tight coupling is a critical part of this, it would potentially be if this was a highly needed feature, but I don't think that there will be a large number of users using this in performance critical areas

tl;dr;
This should, IMO, be a `RefCounted` type with a `HashSet` backing, and hopefully future additions of generics to GDScript can allow us to use typing with them if desired in the future

**Comment by HolonProduction** [comment]
Should we maybe move over to the proposal for further discussions like this?

**Comment by AThousandShips** [comment]
I'd say most of this is implementation related which belongs here, though the broader discussion of `Variant` vs `RefCounted` could be served better over on the proposal

**Comment by AlyceOsbourne** [comment]
I mean, the whole point is all of the set operations.
Yes, its similar to a dict with null values, but half the size because we don't waste a tonne of data, Everything from spatial hashing, to data sieves, to simple deduplication, there are many reasons to have sets in a game engine. 
Every time Sets are bought up in chat, amongst the general users, there is always a display of bafflement to why many of the recent features have been added BEFORE sets, given they are such a fundamental coding tool. 
Within the first week of using Godot I ended up writing a set implementation, because I was having to manually apply set operations, and was genuinely surprised that godot is the only lang I have used that doesn't come with a set. 

In terms of type conversion and coercion. It should be able to swap too and from an array.
For serialization, again, basically an array, with a flag to mark as set.

On the syntax points here, Python has made use of the {} set syntax since forever, and I mean, godot has borrowed a tonne of its syntax from python anyway. 

It seems to me that the argument isn't so much against the inclusion of sets, and more for that fact the type system needs fixing. Its clear that it hasn't been written with the future in mind, and is rigidly designed in such a way that it doesn't support reasonable additions.

**Comment by RadiantUwU** [comment]
> In terms of type conversion and coercion. It should be able to swap too and from an array. For serialization, again, basically an array, with a flag to mark as set.

Serializing will not work if Set is RefCounted, if you try to serialize with `allow_objects` it will serialize it as an `ObjectID`

> On the syntax points here, Python has made use of the {} set syntax since forever, and I mean, godot has borrowed a tonne of its syntax from python anyway.

This is the reason why i wanted to add it using that syntax.

**Comment by dalexeev** [comment]
> I mean, the whole point is all of the set operations.

We could add corresponding methods to `Array` and/or `Dictionary`. For `Array`, we could also take repetition into account, see https://github.com/godotengine/godot-proposals/issues/1756#issuecomment-721152233:

```gdscript
[2, 2, 3].merge_with([2, 5]) # [2, 2, 3, 5]
[2, 2, 3].merge_with([2, 2, 5]) # [2, 2, 3, 5]
[2, 2, 3, 5].intersect_with([2]) # [2]
[2, 2, 3, 5].intersect_with([2, 2]) # [2, 2]
```

> Serializing will not work if Set is RefCounted, if you try to serialize with `allow_objects = false` it will serialize it as an `ObjectID`

We could solve this on the serialization side by adding special handling for the `Set` class. We already support a special syntax for resources in text serialization:

https://github.com/godotengine/godot/blob/f6c7b00a41de9feab5838e07503c2b9fb60d8eb9/core/variant/variant_parser.cpp#L1087

**Comment by HolonProduction** [comment]
> Serializing will not work if Set is RefCounted, if you try to serialize with `allow_objects = false` it will serialize it as an `ObjectID`

That's a good point. I didn't think about that.

> > On the syntax points here, Python has made use of the {} set syntax since forever, and I mean, godot has borrowed a tonne of its syntax from python anyway.
> 
> This is the reason why i wanted to add it using that syntax.

I don't think that really matters. GDScript has the same syntax in a lot of places, but it also borrows from other languages like javascript or lua. In the end we need to evaluate new syntax with GDScript in mind and not python.

> In terms of type conversion and coercion. It should be able to swap too and from an array.

Not sure about that. Rust for example can do all its coercion at compile time, that's why it's basically free for cases were it isn't needed. GDScript being gradual typed can't rely on that. So all code or at least all unsafe code would get slower through this even when not using sets.

**Comment by AThousandShips** [comment]
For array we could also add a method like the STL `unique`, erasing sequential identical elements, to make `Array` more convenient for set operations

**Comment by HolonProduction** [comment]
> We could solve this on the serialization side by adding special handling for the `Set` class. We already support a special syntax for resources in text serialization:
> 
> https://github.com/godotengine/godot/blob/f6c7b00a41de9feab5838e07503c2b9fb60d8eb9/core/variant/variant_parser.cpp#L1087

I guess we don't have a mechanic for modules or gdextensions to add such handlers right?

**Comment by RadiantUwU** [comment]
> > In terms of type conversion and coercion. It should be able to swap too and from an array.
> 
> Not sure about that. Rust for example can do all its coercion at compile time, that's why it's basically free for cases were it isn't needed. GDScript being gradual typed can't rely on that. So all code or at least all unsafe code would get slower through this even when not using sets.

We use a switch case, it doesn't get slower.

**Comment by RadiantUwU** [comment]
I decided to go forward with implementing typed sets as well, as we cannot make RefCounteds currently take container types.

**Comment by adamscott** [comment]
Hi there @RadiantUwU !

I added the `discussion` label because this feature warrants a serious discussion internally about the necessity sets inside `Variant`. I don't want to prejudge, neither predict the outcome of the PR, but there's a lot of chances that it could be rejected on the basis that `Variant` is already stuffed to the brim. Adding types to `Variant` affects heavily performance, as it is used practically everywhere in the engine. So you'll understand why we don't want to "bloat" Variant.

While some of us can offer you reviews and such, it doesn't necessarily mean that this PR would be merged even if everything is alright and up to standards.

You put a lot of work on this PR and we are grateful for it. We'll take time to review it and decide whether or not it is right to merge. We just want you to know the odds of the outcome in advance.

**Comment by peterhoglund** [comment]
@adamscott Any update on this from the team? Have you had the internal discussions?
